fuel-poa-coordinator/src/service.rs

```diff
@@ -0,0 +1,326 @@
+                        self.produce_block().await?;
+                    }
+                    Ok(())
+                }
+            },
+            TxStatus::Executed => Ok(()), // This has been processed already
+            TxStatus::SqueezedOut { .. } => {
+                // TODO: If this is the only tx, set timer deadline to last_block_time + max_block_time
+                Ok(())
+            }
+        }
+    }
+
+    async fn on_timer(&mut self, _at: Instant) -> anyhow::Result<()> {
+        match self.trigger {
+            Trigger::Instant | Trigger::Never => {
+                unreachable!("Timer is never set in this mode");
+            }
+            // In the Interval mode the timer expires only when a new block should be created.
+            // In the Hybrid mode the timer can be either:
+            // 1. min_block_time expired after it was set when a block
+            //    would have been produced too soon
+            // 2. max_tx_idle_time expired after a tx has arrived
+            // 3. max_block_time expired
+            // => we produce a new block in any case
+            Trigger::Interval { .. } | Trigger::Hybrid { .. } => {
+                self.produce_block().await?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Processes the next incoming event. Called by the main event loop.
+    /// Returns Ok(false) if the event loop should stop.
+    async fn process_next_event(&mut self) -> anyhow::Result<bool> {
+        tokio::select! {
+            _ = self.stop.recv() => {
+                Ok(false)
+            }
+            txpool_event = self.txpool_broadcast.recv() => {
+                self.on_txpool_event(&txpool_event.context("Broadcast receive error")?).await.context("While processing txpool event")?;
+                Ok(true)
+            }
+            at = self.timer.wait() => {
+                self.on_timer(at).await.context("While processing timer event")?;
+                Ok(true)
+            }
+        }
+    }
+
+    async fn init_timers(&mut self) {
+        match self.trigger {
+            Trigger::Never | Trigger::Instant => {}
+            Trigger::Interval { block_time } => {
+                self.timer
+                    .set_timeout(block_time, OnConflict::Overwrite)
+                    .await;
+            }
+            Trigger::Hybrid { max_block_time, .. } => {
+                self.timer
+                    .set_timeout(max_block_time, OnConflict::Overwrite)
+                    .await;
+            }
+        }
+    }
+
+    /// Start event loop
+    async fn run(mut self) {
+        self.init_timers().await;
+        loop {
+            match self.process_next_event().await {
+                Ok(should_continue) => {
+                    if !should_continue {
+                        break
+                    }
+                }
+                Err(err) => {
+                    // TODO: is this the right way to handle an error here?
```

Comment:
Error processing I think should be fleshed out more here. If `process_next_event` errs with a txpool error it might be reasonable to dump the txpool and try to re-populate it, while if there is an error with block production skipping the block or some other recourse like changing block production strategy may be a more desireable outcome. So maybe seperate out the err cases within here to catch both of those, then as a final exhaustive match panic

Reply:
I'm not exactly sure what do you mean by repopulating the txpool here? I don't think we even do any fallible txpool calls anyway. My reasoning is that if we end up taking the error path here, it must already be a programming error somewhere. I would like to panic early instead of doing this, since it's hightly unlikely that any error can be recovered from.